doc/developer/design/20220226_multiwriter_persist.md

```diff
@@ -0,0 +1,206 @@
+that are contiguous in time.
+
+If a collection is being written to by multiple writers, and therefore there
+are multiple pending areas, then the batches of the two areas must contain
+identical updates for all times `t`. If this invariant is not upheld then the
+contents of the persistent collection are undefined.
+
+Note that there is no requirement that the `lower` and `upper` frontiers of the
+batches in two separate pending areas to match up. Only that the diffs at each
+particular time are identical.
+
+## Collection lifecycle
+
+A persisted collection is managed by the PERSIST CONTROLLER which has the following API:
+
+* `CreateCollection(id)` -> ReadHandle: Creates a new persistent collection
+  identified by `id` and returns a read handle. After a persistent collection
+  has been created anyone can publish new batches into its pending area.
+* `AcquireReadHandle(id) -> ReadHandle`: provides a read handle at its current
+  `since` frontier and referring at the latest merged area version.
+* `DeleteCollection(id)`: Deletes a persistent collection identified by `id`
+
+### Reading
+
+A read handle of a persistent collection is described by a merged area version
+and a compaction frontier. **Readers of a collection must use the batches
+referred to by their current merged area version for any reads at times not
+beyond the merged area's `upper`. For times beyond `upper` the pending areas
+are used instead.**
+
+This behavior is the key to allow the PERSIST CONTROLLER to decide with
+certainty which batches are reachable and perform background garbage collection
+of unreachable batches.
+
+When a reader wants to read at a time `t` that is beyond the merged area's
+`upper` there might be multiple pending areas (due to multiple writers)
+containing a batch that includes that time. Readers are free to choose which
+pending area to read from based on any strategy. For example, a reader could
+pick a pending area to follow and only switch to a different one if its current
+pending area is falling behind too much. The trade-off for switching which
+pending area is followed is the additional work that needs to be done if the
+frontiers of the batches don't exactly line up.
+
+Readers continually try to discover newly minted merged area versions and
+increase their associated version to that, indicating that they will never use
+the batch metadata of any previous version.
+
+Readers are also able to downgrade their compaction frontier indicating that
+they no longer care to distinguish updates that happened not beyond the
+frontier.
+
+Both compaction frontier downgrades and merged area version downgrades are
+durably persisted.
+
+A read handle can be created by cloning an existing read handle. The newly
+created read handle will have its own independent merged area version and
+compaction frontier with their initial values set to the values of the read
+handle it was cloned from. This is also a durable operation.
+
+### Writing
```

Comment:
How does the PERSIST CONTROLLER know when a writer has stopped writing? Is the writer responsible for deleting its pending area on exit / somehow notifying the persist controller that its no longer going to exist?

Concretely:

two writers w1 and w2 start up. w1 writes up to `t1` in its pending area and then exits. Now `w2` writes up to `t2`. We'd like to move the writes at t2 into the merged area, but if we don't know that `w1` has exited, do we have to wait for `w1` to write to `t2` before moving things into the merged area? Alternatively, do we assume that the invariant that all writers write the same (data, time, diff)s holds and then move the writes at t2 into merged, even though w1 has not said anything yet?

Reply:
I guess we could decouple "moving things into merged" from "validating the invariant" where the former can happen opportunistically, but the latter waits for all writers to write to a given time

Reply:
In general the tradeoff of switching writers is that the batch `lower` `upper` frontiers might not match up exactly and so there is extra work involved to split a batch into two parts. Similarly to readers the compaction process could wait for w1 to produce its correctly aligned batch up to some timeout after which it can decide to switch to `w2` and start following that stream

Reply:
Right. it makes sense that the compacter could listen to only one worker and switch around like listeners, but then I think the compacter is no longer able to check that the invariant "all writers wrote the same (data, time, diff)" is true. I was trying to think about what needs to be possible in order for that invariant to be checkable